### ISSUE TITLE
lseg

### ISSUE THREAD

**Issue by mrocr** [ocropus/ocropus4#7]
@tmbdev Hi there,

i have [noticed](http://storage.googleapis.com/ocropus4-test/gsub-lseg-test.tar) that you are testing the segmentation of text-lines, you seem to have chosen the classes of:
(x-height) + (textline boundary) + (text region)

It's interesting to hear about your results:
- How does it perform against **connected components**, example of 2 lines close or touching each other.
- Can it **generalize** to unseen documents with different fonts and noise.
- Can it segment vertical and curved lines

by-the-way, i have noticed that some labels in the `gsub-lseg-test.tar` might need correction, an example is attached
[sample.zip](https://github.com/ocropus/ocropus4/files/7069095/sample.zip)

**Comment by tmbdev** [comment]
By default, OCRopus 4 uses word segmentation and word recognition, not line segmentation.

The published training data is automatically derived using other OCR engines and contains lots of errors. They tend to go away after multiple rounds of EM (semi-supervised/self-supervised training).

For both word and line segmentation, there are three classes: geometric center (nucleus), periphery, and background. Each nucleus determines a text line. For each nucleus, only pixels in the periphery are included. If a connected periphery contains only one nucleus, then all pixels are assigned to the word representing that nucleus. If a connected periphery contains multiple nuclei, pixels in the periphery are assigned to the nearest nucleus. 

The algorithm works reasonably well with touching text lines. Given the training data we have, the segmentation isn't precise, but it's good enough for recognition. With better training data, the same framework could be used for pixel-exact segmentation.

The algorithm doesn't care about orientation; if you train it on vertical or curved text, it will segment that kind of text just fine. We have used it for scene text as well.

It seems to generalize quite well to different classes of documents; that mainly depends on available training data and augmentations used during training. By default, we train with UW3, G1000, Tobacco Corpus, and public documents from the Archive. There's a lot more to be done to optimize the training set composition and the augmentations.

**Comment by mrocr** [comment]
It's great to see that you have found a cross-point between biomedical and document image segmentation.

My intuition:
- i anticipate that we need no more than 3 masks to detect both printed and handwritten text with high accuracy,
one-of-which is either a base-line or x-height.
- smooth blending of patches can be used to maintain high accuracy with low memory.
- For printed text, the dataset doesn't have to be superb, as it might be better to focus on adding many different augmentations, such as decreasing or increasing the gaps between texlines or spaces between words, etc...

Conclusion:
A printed version can be released currently, while in the future a universal version for both handwriting and printed can be released by simply using a pixel-wise dataset, along with modifying the periphery mask to include some form of a weighted loss in the mask.
